### PR TITLE
fix(win32): show a tab's whole title without renaming the tab

### PR DESCRIPTION
**Pull request by nanasess**
## Problem

A tab whose title outgrows its button had no way to show the rest, so renaming
the tab was the only way to read it back. Two things combined:

- `compactHostLabel` compacted the label against a **byte** budget. The widest
  tab gets 24 of them, so a Japanese title lost everything past the eighth
  character; a narrow tab showed two.
- Nothing else in the chrome carried the whole string. There was no tooltip, and
  `tabItemUiaName` reported the same truncated label to assistive technology.

## What upstream does

Upstream Ghostty never has to place this popup itself:

- **GTK** — `src/apprt/gtk/ui/1.5/tab.blp` binds `title` straight to
  `AdwTabPage`. libadwaita fades the label to the pixel width, and `adw-tab.c`'s
  `update_tooltip` falls back to the page title when the page sets none of its
  own.
- **macOS** — native window tabs; the SwiftUI title is `.lineLimit(1)` +
  `.truncationMode(.tail)`.

Neither counts bytes or codepoints. This brings the Win32 chrome to the same
place with the pieces it has to own itself.

## Changes

**A cell budget instead of a byte budget.** `compactHostLabel` now measures with
the same width table the terminal grid uses, so the budget its callers derive
from a pixel width is spent in the unit a pixel width actually buys. ASCII
labels are unchanged; a CJK title gets the 12 characters its 24 cells are worth
instead of 8. A double-width codepoint that would straddle the budget is dropped
rather than allowed to spill, and the cut still lands on a UTF-8 boundary, which
`utf8ToUtf16LeAllocZ` requires. `compactHostLabelWidth`, `hostTabLabelMaxWidth`,
`profileStatusBadgeTextWidth` and `tab_label_max_width` are renamed with it
because their unit changed — the launcher chip reserved `len * 7` px from the
byte count and over-reserved for every non-ASCII label.

**A hover tooltip** for tabs whose drawn label had to be compacted, in
`win32_tab_tooltip.zig` (placement math, unit-tested) plus the window itself. It
is a top-level `WS_POPUP` owned by the host, armed off `TME_HOVER` so it waits
out the system hover delay. Top level is not incidental: as a child window it
drew correctly and was then painted away by the terminal surface's next frame,
because the surfaces are siblings repainting on their own clock and no Z order
among children survives that. `place` keeps the popup inside the host's client
area in both axes — nothing else would clip a top-level window — and the popup's
own `DT_END_ELLIPSIS` truncates a title too wide for that. It is
hit-transparent so a pointer near the tab's bottom edge cannot flicker it, and
it hides on click, drag, leave, label rebuild and window deactivation — a
pointer that never left the tab sends no `WM_MOUSELEAVE`, so losing the window
has to take it down explicitly. A tab whose label is drawn whole gets no
tooltip.

The popup carries its own `ChromeControlProvider`, per AGENTS.md's rule that a
PR adding a widget adds its provider. `ChromeControlState.Role` gains `tooltip`,
reporting `UIA_ToolTipControlTypeId`; the provider's name is the full title and
it raises a name-changed event when the popup moves to another tab.

**`tabItemUiaName`** reports `N: <full title>` instead of the drawn label, and
falls back to that label only when the title does not fit the caller's buffer.

The three are one change: the same title was unreachable to the eye, to the
pointer and to a screen reader, and the tooltip's "is anything hidden?" test is
the cell measurement the first part adds.

## Validation

- `zig build -Doptimize=ReleaseSafe -Demit-exe=true` → `Build Summary: 75/75
  steps succeeded`
- `zig build test` over the label, tooltip, tab-label and UIA filters plus the
  `ConPTY` root test → `214/218 tests passed; 4 skipped`, 0 failed. 10 tests are
  new. I confirmed they actually execute by breaking one in each new file and
  watching exactly those two fail (`expected 3, found 2` and `expected 41, found
  40`), then reverting.
- `test/windows/flagship/Test-VerificationContracts.ps1` → `flagship
  verification contracts: PASS (2 scenarios)`.
- Manual, Windows 11 24H2, `--single-instance=false`, ReleaseSafe build:
  - A long Japanese title now fills the tab label instead of stopping at 8
    characters.
  - The tooltip holds for as long as the pointer rests on the tab.
  - No tooltip for a title that already fits.
  - The right-most tab's tooltip stays inside the window.
  - No flicker with the pointer near the tab's bottom edge.
  - Same results under the default configuration
    (`--config-default-files=false`) in both `--window-theme=dark` and
    `--window-theme=light`.
  - After the review fixes: a tooltip appears once enough tabs are open to
    squeeze the labels, the popup stays inside a narrowed window, and short
    titles still get nothing until the tab is narrow enough to ellipsize them.

## Risks and follow-up

- The cell budget still comes from an average glyph advance
  (`(button_width - 26) / 8`), so it stays an estimate. GDI's `DT_END_ELLIPSIS`
  remains the backstop, exactly as before this change.
- A title made only of zero-width codepoints measures 0 cells, is never
  compacted, and so reaches the label whole for GDI to clip. Nothing overruns:
  the tooltip's text buffer is fixed and `GetWindowTextW` truncates into it.
- The tooltip hides on window deactivation but does not follow a window move or
  resize while it is up. Reaching that needs a pointer parked on a tab while the
  window is dragged by other means; the next hover replaces the popup.
- Clamping the popup to the client area means a title wider than the window is
  itself ellipsized in the tooltip. A popup drawn past the window would have put
  that text over the neighbouring app, or off-screen where it is unreachable
  anyway, so the clamp is the better of the two failures.

## Review follow-ups

Three of the four bot findings were real and are fixed in the third commit:

- The tooltip check measured the bare title, but the drawn label also carries
  the active marker, the 1-based index and the pane count, and GDI ellipsizes
  all of it. A title that fit on its own could still lose its tail with no
  tooltip. It now measures the drawn label.
- `place` clamped only the left edge. That was right while the popup was a
  child window and the parent clipped it; it was wrong once the popup went top
  level. The width is clamped too.
- The popup had no UIA provider, which AGENTS.md requires for a new widget.

The fourth said a `?usize` field compared against a `u32` `Tab.id` cannot
compile. It does — Zig widens implicitly, and the branch is green — but the
field is `?u32` now, since it had no reason to be wider than what it stores.

## AI assistance

Written with Claude Code (Opus 5), per AI_POLICY.md. I reviewed the change,
verified the upstream GTK/macOS behaviour it is modelled on against
`ghostty-org/ghostty` at `492300cad`, and ran the validation above myself.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by Sourcery

Make truncated Win32 tab titles readable through cell-aware labels, hover tooltips, and complete accessibility names.

New Features:
- Add hover tooltips that reveal the full title of truncated Win32 tabs while keeping the popup within the host window.
- Expose full tab titles to UI Automation clients, including an accessible tooltip control.

Bug Fixes:
- Measure Win32 tab and profile labels in display cells rather than bytes, preserving more non-ASCII text and avoiding invalid UTF-8 truncation.

Enhancements:
- Coordinate tooltip lifecycle with tab interaction, relayout, window activation, and dragging to prevent stale or flickering overlays.
- Add tested tooltip placement behavior for edge and constrained-window cases.

Tests:
- Expand Win32 label coverage for display widths, wide characters, malformed bytes, truncation detection, tooltip bounds, and accessible title behavior.

## Summary by CodeRabbit

- **New Features**
  - Hover over a shortened tab label to view its full title in a tooltip.
  - Tooltips automatically position themselves to remain visible within the window.
  - Accessibility information now includes each tab’s index and complete title.

- **Improvements**
  - Tab, profile, and command-preview labels now account for display width, improving truncation for wide characters such as CJK text.
  - Text truncation avoids cutting characters inappropriately.
  - Tooltip text is bounded to keep it readable and unobtrusive.

### CODE REVIEW

File: src/apprt/win32.zig [hunk 1]
```diff
@@ -13635,6 +13670,9 @@ const Host = struct {
     }
 
     fn destroyTabButton(self: *Host, tab: *Tab) void {
+        if (self.tab_tooltip_tab_id) |shown| {
+            if (shown == tab.id) self.hideTabTooltip();
+        }
```
[the remark this answers is not in the document]
Comment (nanasess) [reply src/apprt/win32.zig:13677]: The branch compiles: `Windows Build And Tests` passed on f5472fef1, and `zig build -Doptimize=ReleaseSafe -Demit-exe=true` is green locally. Zig widens `u32` to `usize` implicitly, and the `==` comparison resolves through peer type resolution, so neither the assignment nor the comparison is an error.

That said, the field had no reason to be wider than the value it stores. It is `?u32` as of 566250743, matching `Tab.id`, so nobody has to reason about the coercion to read it.

File: src/apprt/win32.zig [hunk 2]
```diff
@@ -13691,6 +13729,178 @@ const Host = struct {
...
+        const index = self.tabIndexForButton(button) orelse return self.hideTabTooltip();
+        const tab = &self.tabs.items[index];
+        if (self.tab_tooltip_tab_id) |shown| {
+            // The hover timer re-arms on every pointer move inside the tab, so
+            // this fires repeatedly while the tooltip is already up.
+            if (shown == tab.id) return;
+        }
+        const title = tab.cached_button_title orelse return self.hideTabTooltip();
+        if (!labels.hostLabelIsCompacted(title, tab.cached_button_label_max_width)) {
+            return self.hideTabTooltip();
```
[the remark this answers is not in the document]
Comment (nanasess) [reply src/apprt/win32.zig:13765]: Correct, and thanks for catching it. The check compared the bare title, but `buildTabButtonLabel` prepends the active marker and the 1-based index and can append the pane count, and GDI applies `DT_END_ELLIPSIS` to all of it inside a button that has already given up its close zone. A title that fit its own budget could still lose its tail with no tooltip to recover it.

Fixed in 566250743: the check now measures `cached_button_label` — what is actually drawn — against the same cell budget.

File: src/apprt/win32_tab_tooltip.zig [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+/// the right edge slides left instead of hanging off the window; the last
+/// resort is the left margin, where a title wider than the window is clipped by
+/// the popup's own `DT_END_ELLIPSIS` rather than drawn out of view.
+pub fn place(anchor: Rect, size: Size, client: Rect, gap: i32, margin: i32) Rect {
+    const w = @max(0, size.width);
+    const h = @max(0, size.height);
+
+    var left = anchor.left;
+    if (left + w > client.right - margin) left = client.right - margin - w;
+    if (left < client.left + margin) left = client.left + margin;
```
[the remark this answers is not in the document]
Comment (nanasess) [reply src/apprt/win32_tab_tooltip.zig:54]: Agreed. Clamping only the left edge was defensible while the popup was a `WS_CHILD` and the parent clipped it; it stopped being defensible when the popup became a top-level `WS_POPUP`, where nothing clips it and anything past the screen edge is unreachable anyway.

Fixed in 566250743: `place` clamps the width to the client area minus both margins, and the popup's own `DT_END_ELLIPSIS` truncates what will not fit. The test that pinned the old behaviour now asserts the popup stays inside `client.right - margin`.

File: src/apprt/win32.zig [hunk 4]
```diff
@@ -13691,6 +13729,178 @@ const Host = struct {
...
+                0,
+                0,
+                0,
+                host_hwnd,
+                null,
+                self.app.hinstance,
+                null,
+            ) orelse return;
+            setWindowData(created, self);
+            self.tab_tooltip_hwnd = created;
```
[the remark this answers is not in the document]
Comment (nanasess) [reply src/apprt/win32.zig:13829]: Correct — AGENTS.md is explicit that a PR adding a UI widget adds its provider, and exposing the title through the tab item does not describe the popup itself.

Fixed in 566250743: `ChromeControlState.Role` gains `tooltip`, which reports `UIA_ToolTipControlTypeId` (50042), and the popup window carries a `ChromeControlProvider` whose name is the full title. It raises a name-changed event when the popup moves to a different tab, and is detached alongside the window on teardown. It exposes no pattern and is not keyboard focusable, matching what the widget actually is.

File: src/apprt/win32_tab_tooltip.zig [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+/// Place a tooltip of `size` for the tab button occupying `anchor`, keeping the
+/// popup inside `client` with `margin` px to spare.
+///
+/// The tooltip hangs below its tab by `gap` px and is left-aligned with it, so
+/// the eye can follow the label straight down into the full title. A tab near
+/// the right edge slides left instead of hanging off the window; the last
+/// resort is the left margin, where a title wider than the window is clipped by
+/// the popup's own `DT_END_ELLIPSIS` rather than drawn out of view.
+pub fn place(anchor: Rect, size: Size, client: Rect, gap: i32, margin: i32) Rect {
+    const w = @max(0, size.width);
```
[the remark this answers is not in the document]
Comment (nanasess) [reply src/apprt/win32_tab_tooltip.zig]: Fixed in 566250743, as suggested: `place` clamps the width to `client.width() - margin * 2` before it resolves `left`, so the popup can no longer extend past the host. The text that does not fit is truncated by the popup's own `DT_END_ELLIPSIS`.

The clamp matters more than it looks: the popup is a top-level `WS_POPUP`, so before this it could be drawn over whatever sat beside the window.

**Comment by amanthanvi** [comment]
Thanks @nanasess — merging this. The cell-budget change, the top-level tooltip popup, and the UIA name fix are all sound, and the three bot findings you addressed in 566250743 check out.

Before merging I ran an independent review pass on top of the bot reviews and re-verified locally: `zig build -Demit-exe=true`, the label and tooltip test filters (all ten new tests execute), `Test-VerificationContracts.ps1`, and `zig fmt --check`. Two small things came out of that which I'll handle in a maintainer follow-up rather than hold this PR for:

- `hostLabelIsCompacted` measures the decorated label (`* `, `N: `, ` (panes)`) against the budget the bare title was compacted to, so a title that fits can still trip the tooltip on the active tab and not on its neighbour. Subtracting the decoration width before compacting fixes both that and the drawn label overflowing the button by up to 9 cells.
- `paintTabTooltip` can return before `BeginPaint` if `tab_tooltip_hwnd` is null, which would leave `WM_PAINT` unvalidated. Unreachable today, but cheap to harden.

---
*This comment was written by an AI agent (Claude Code) operating the maintainer's account, with the maintainer's authorization. The review findings above were verified against the code and local test runs.*
